### ISSUE TITLE
Reading justfiles from stdin

### ISSUE THREAD

**Issue by gl-yziquel** [casey/just#2206]
Hi.

As I'd like to declare just fragments remotely and get them executed locally, I'd like to be able to fetch them via curl and get them piped into just. Doesn't seem to currently be possible.

Something like just --sdtin would be nice.

**Comment by laniakea64** [comment]
It is possible:
```
$ echo 'foo:
>   echo foo
> 
> bar:
>   echo bar' | just -f /dev/stdin
echo foo
foo
```

But note that this method uses `/dev` as working directory:
```
$ echo 'a:
>   pwd' | just -f /dev/stdin
pwd
/dev
```

There is https://github.com/casey/just/pull/1933 for adding proper support

**Comment by gl-yziquel** [comment]
@laniakea64 I thought about this. Unfortunately, I really do not like the idea of having the folder switched to `/dev` when using `/dev/stdin` as a Justfile...

For now, I worked around it with a temporary file. But I also do not like it at all.

**Comment by laniakea64** [comment]
> I really do not like the idea of having the folder switched to `/dev` when using `/dev/stdin` as a Justfile...

Yeah, me neither.  Although there is also this option as a workaround:
```
$ echo 'a:
>   pwd' | just -d . -f /dev/stdin
```

**Comment by gl-yziquel** [comment]
> > I really do not like the idea of having the folder switched to `/dev` when using `/dev/stdin` as a Justfile...
> 
> Yeah, me neither. Although there is also this option as a workaround:
> 
> ```
> $ echo 'a:
> >   pwd' | just -d . -f /dev/stdin
> ```

Pretty cool.

**Comment by casey** [comment]
`/dev/stdin` is a great workaround, but it doesn't work on Windows, and it should probably be easier, so I'm going to reopen this.

#1933 is currently open to implement this.

**Comment by WieeRd** [comment]
Just discovered this issue while thinking about the same feature.
`curl | just -` could be an interesting alternative to the typical `curl | sh` setup scripts.

I have one question/concern about the implementation: How would it deal with `just` calls inside the recipes?

```just
default:
    just foo
    just bar

foo:
    ...

bar:
    ...
```

Dependencies (`default: foo bar`) would work fine even when the Justfile is fed through stdin. But invoking `just recipe` command within the recipe is also a common practice. Those are simply shell commands that call another `just` process expecting a Justfile to be in the current working directory, right? And there won't be one in the case of reading Justfile from stdin. What kind of special logic would be required to make these keep functioning?

**Comment by casey** [comment]
Implemented in #3325. It's a little gross, since it writes the justfile to a temporary directory and runs it from there. However, doing it the "right" and reading the source from memory would have been a much larger diff, and why I put this off for so long. If this approach causes problems we can deal with it when they come up. The working directory is the invocation directory, if no working directory is given.
